src/mgr/BaseMgrModule.cc

```diff
@@ -819,19 +829,80 @@ ceph_add_osd_perf_query(BaseMgrModule *self, PyObject *args)
         }
         query.performance_counter_descriptors.push_back(it->second);
       }
+    } else if (query_param_name == NAME_LIMIT) {
+      if (!PyDict_Check(query_param_val)) {
+        derr << __func__ << " query " << query_param_name << " not a dict"
+             << dendl;
+        Py_RETURN_NONE;
+      }
+
+      *limit = {};
+      PyObject *limit_params = PyDict_Items(query_param_val);
+
+      for (int j = 0; j < PyList_Size(limit_params); ++j) {
+        PyObject *kv = PyList_GET_ITEM(limit_params, j);
+        char *limit_param_name = nullptr;
+        PyObject *limit_param_val = nullptr;
+        if (!PyArg_ParseTuple(kv, "sO:pair", &limit_param_name,
+                              &limit_param_val)) {
+          derr << __func__ << " limit item " << j << " not a size 2 tuple"
+               << dendl;
+          Py_RETURN_NONE;
+        }
+
+        if (limit_param_name == NAME_LIMIT_ORDER_BY) {
+          if (!PyString_Check(limit_param_val)) {
+            derr << __func__ << " " << limit_param_name << " not a string"
+                 << dendl;
+            Py_RETURN_NONE;
+          }
+          auto order_by = PyString_AsString(limit_param_val);
+          auto it = counter_types.find(order_by);
+          if (it == counter_types.end()) {
+            derr << __func__ << " limit " << limit_param_name
+                 << " not a valid counter type" << dendl;
+            Py_RETURN_NONE;
+          }
+          limit->order_by = it->second;
+        } else if (limit_param_name == NAME_LIMIT_MAX_COUNT) {
+#if PY_MAJOR_VERSION <= 2
+          if (!PyInt_Check(limit_param_val) && !PyLong_Check(limit_param_val)) {
+#else
+          if (!PyLong_Check(limit_param_val)) {
+#endif
+            derr << __func__ << " " << limit_param_name << " not an int"
+                 << dendl;
+            Py_RETURN_NONE;
+          }
+          limit->max_count = PyLong_AsLong(limit_param_val);
```

Comment:
Does this work under Python2 if the type is an ```int``` or do you have to use ```PyInt_AsLong```?

Reply:
It works in my python2 env. In my case `PyInt_Check(limit_param_val)` returns true and `PyLong_AsLong(limit_param_val)` does correct conversion. `PyInt_AsLong(limit_param_val)` works for me to, so we can change this to `PyInt_AsLong` (for Python3 it still will call PyLong_AsLong according to mgr/PythonCompat.h).

I am not sure though for the `PyLong_Check(limit_param_val)` case for the Python2, how it is supposed to be triggered. And if it is eventually triggered for someone, will `PyInt_AsLong` work?

Reply:
OK, if it works it works. I was thinking you might need to special-case it to "if it's an int, use PyInt_ToLong else if it's a long use PyLong_ToLong".